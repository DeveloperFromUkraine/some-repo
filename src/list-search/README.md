## API
    ### Selector
        * ign-list-search

    ### Exported As
        * ListSearchComponent
    
    ### Properties
    * `@Input() ariaLabel: string`
        * Takes in descriptive label to set aria-label to
    * `@Output() searchCriteria$: BehaviorSubject`
        * Outputs search values for consumers to subscribe to

## Implementation HTML
    <ign-list-search ariaLabel="Search employees"></ign-list-search>

## Implementation TS
    None
