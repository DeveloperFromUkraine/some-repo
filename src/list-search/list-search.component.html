<mat-icon
  class="search-icon">search</mat-icon>
<mat-form-field
  floatPlaceholder="never">
  <input
    matInput
    placeholder="Search"
    attr.aria-label="{{ariaLabel | translateIDS}}"
    #searchInput
    (keyup.escape)="searchInput.value=''; emitSearchEvents()"
    (keyup)="emitSearchEvents()">
  <button
    *ngIf="searchInput.value"
    aria-label="Clear"
    mat-icon-button
    matSuffix
    (click)="searchInput.value=''; emitSearchEvents()">
    <mat-icon>close</mat-icon>
  </button>
</mat-form-field>
