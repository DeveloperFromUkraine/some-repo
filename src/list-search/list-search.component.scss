mat-icon {
  color: rgba(0,0,0,.54);

  &.search-icon {
    width: 10%;
    margin: 0 0.5rem 0.25rem 0;
  }
}

mat-form-field {
  font-size: 1rem;
  width: 90%;
}

:host {
  display: flex;
  align-items: center;
  width: 100%;
}
