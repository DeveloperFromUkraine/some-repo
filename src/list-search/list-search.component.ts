import { Component, ElementRef, Input, Output, ViewChild, ChangeDetectorRef } from '@angular/core';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { TranslationPipe } from '../localization/translation.pipe';

@Component({
  selector: 'ign-list-search',
  templateUrl: './list-search.component.html',
  styleUrls: ['./list-search.component.scss'],
})
export class ListSearchComponent {
  @ViewChild('searchInput') searchInput: ElementRef;

  @Input() ariaLabel;
  @Output() searchCriteria$ = new BehaviorSubject<any>({});

  constructor(private _cdr: ChangeDetectorRef) {
    const pipe = new TranslationPipe(this._cdr);
    this.ariaLabel = pipe.transform('LIST_SEARCH');
  }

  emitSearchEvents() {
    this.searchCriteria$.next({
      searchBy: this.searchInput.nativeElement.value,
    });
  }
}
