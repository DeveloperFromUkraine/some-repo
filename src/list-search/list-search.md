*** HTML ***
# List Search

## Use
A search component which displays a search icon and an input field to be used with lists.

## Sample

<mat-tab-group>
    <mat-tab label="Component Sample"><div class="tab-height">
        <ign-list-search ariaLabel="Search employees"></ign-list-search>
    </div></mat-tab>
    <mat-tab label="HTML"><div class="tab-height">
        <table style="width:100%">
            &lt;ign-list-search ariaLabel="Search employees"&gt;&lt;/ign-list-search&gt;
         </table></div>
    </mat-tab>
</mat-tab-group>

## Content
* `@Input() ariaLabel: string`
    * Takes in descriptive label to set aria-label to
* `@Output() searchCriteria$: BehaviorSubject`
    * Outputs search values for consumers to subscribe to


## Style

* The list search component can be used at the top of a list by itself or accompanied by other options in the row, such as a filter button.

## Accessibility

* Add a descriptive ariaLabel attribute that describes what is being searched (for example: "Search Employees")
* Refer to Sample HTML for an example on usage

*** TypeScript *** 

*** ExternalTypeScript ***

*** Imports ***
