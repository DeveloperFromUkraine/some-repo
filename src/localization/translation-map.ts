export const TranslationMap = {
  CLEAR: 'Clear',
  CLOSE: 'Close',
  CONTEXT_MENU: 'Context Menu',
  EDIT: 'Edit',
  END_DATE_TYPEAHEAD: 'End Date (Typeahead entries appear in dropdown)',
  JUMP_TO_CONTENT: 'Jump to content',
  OPEN: 'Open',
  LIST_SEARCH: 'List Search',
  LOADING: 'Loading',
  SEARCH: 'Search',
  START_DATE_TYPEAHEAD: 'Start Date (Typeahead entries appear in dropdown)',
};
