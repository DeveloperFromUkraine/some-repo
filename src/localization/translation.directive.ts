import { Directive, ElementRef, AfterViewInit } from '@angular/core';
import { TranslationService } from './translation.service';

@Directive({
  selector: '[translateIDS]'
})
export class TranslationDirective  implements AfterViewInit {

    constructor(private element: ElementRef) {
        // Subscribe to language updates
        TranslationService.changeTranslationObservable.subscribe(translation => {
            this.updateTranslation(translation)
        })
    }

    ngAfterViewInit() {
        // set the translation of all content at start
        this.updateTranslation(TranslationService.getTranslation());
    }

    updateTranslation(translation: any) {
        let nodes: NodeList = this.element.nativeElement.childNodes;
        for (let i = 0; i < nodes.length; ++i) {
            let node: any = nodes[i];
            if (node.nodeType === 3) { // node type 3 is a text node
                if (node.originalContent === undefined) { // we store the translation token in original content of node
                    node.originalContent = this.getContent(node);
                }

                const translateValue = this.getTranslation(node.originalContent.trim(), translation);
                this.setContent(node, translateValue);
            }
        }
    }

    getTranslation(translationKey: string, translation: any) {
        return translation[translationKey];
    }

    getContent(node: any) {
        return node.data;
    }

    setContent(node: any, value: string) {
        node.data = value;
    }
}
