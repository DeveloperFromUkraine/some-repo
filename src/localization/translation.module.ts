import { NgModule } from '@angular/core';
import { TranslationPipe } from './translation.pipe';
import { TranslationDirective } from './translation.directive';

@NgModule({
    imports: [
    ],
    declarations: [
        TranslationPipe,
        TranslationDirective
    ],
    exports: [
        TranslationPipe,
        TranslationDirective
    ]
})
export class TranslationModule {}
