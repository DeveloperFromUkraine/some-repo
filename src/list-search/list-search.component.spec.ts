import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ListSearchComponent } from './list-search.component';
import { ComponentTest } from '../../test/test-bed/component';
import { By } from '@angular/platform-browser';
import { DebugElement } from '@angular/core/src/debug/debug_node';
import { Component, NgModule } from '@angular/core';
import { TranslationModule } from '../localization/translation.module';
import { TranslationMap } from '../localization/translation-map';

describe('List Search Component', () => {
  let fixture: ComponentFixture<ListSearchComponent>;
  let component: ListSearchComponent;
  let de: DebugElement;
  let ne: HTMLElement;

  beforeEach(() => {
    ComponentTest.createTestBed(
      [TranslationModule] as NgModule[],
      [ListSearchComponent] as Component[]
    );

    fixture = TestBed.createComponent(ListSearchComponent);
    component = fixture.componentInstance;

    fixture.detectChanges();
  });
  describe('Input field', () => {
    beforeEach(() => {
      de = fixture.debugElement.query(By.css('input'));
      ne = de.nativeElement;
    });
    describe('On key up', () => {
      it('Should call emitSearchEvents', () => {
        jest.spyOn(component, 'emitSearchEvents');

        de.triggerEventHandler('keyup', null);

        expect(component.emitSearchEvents).toHaveBeenCalledTimes(1);
      });

      it('Should emit searchEvent', () => {
        let spy = jest.fn();

        component.searchCriteria$.subscribe(spy);
        de.triggerEventHandler('keyup', null);

        expect(spy).toHaveBeenCalled();
      });
    });
  });
  describe('Clear button', () => {
    beforeEach(() => {
      component.searchInput.nativeElement.value = 'test';

      fixture.detectChanges();

      de = fixture.debugElement.query(By.css('button'));
      ne = de.nativeElement;
    });
    it('Should display when value present', () => {
      expect(ne).not.toBeNull();
    });
    describe('When clicked', () => {
      it('Should remove value', () => {
        ne.click();

        fixture.detectChanges();

        expect(component.searchInput.nativeElement.value).toBe('');
      });
    });
    it('Should call emitSearchEvents', () => {
      jest.spyOn(component, 'emitSearchEvents');

      de.triggerEventHandler('click', null);

      expect(component.emitSearchEvents).toHaveBeenCalledTimes(1);
    });

    it('Should emit searchEvent', () => {
      let spy = jest.fn();

      component.searchCriteria$.subscribe(spy);
      de.triggerEventHandler('click', null);

      expect(spy).toHaveBeenCalled();
    });
  });
  describe('aria-label', () => {
    beforeEach(() => {
      de = fixture.debugElement.query(By.css('input'));
      ne = de.nativeElement;
    });
    it('Should be set to passed in value', () => {
      component.ariaLabel = 'Search employees';

      fixture.detectChanges();

      expect(ne.getAttribute('aria-label')).toBe('Search employees');
    });
    it('Should be set to the default value when no value is passed in', () => {
      fixture.detectChanges();

      expect(ne.getAttribute('aria-label')).toBe(TranslationMap['LIST_SEARCH']);
    });
  });
  describe('On escape key press', () => {
    beforeEach(() => {
      de = fixture.debugElement.query(By.css('input'));
      ne = de.nativeElement;
    });
    it('Should set value to an empty string', () => {
      de.triggerEventHandler('keyup.escape', null);

      fixture.detectChanges();

      expect(component.searchInput.nativeElement.value).toBe('');
    });
    it('Should call emitSearchEvents', () => {
      jest.spyOn(component, 'emitSearchEvents');

      de.triggerEventHandler('keyup.escape', null);

      expect(component.emitSearchEvents).toHaveBeenCalledTimes(1);
    });
    it('Should emit searchEvent', () => {
      let spy = jest.fn();

      component.searchCriteria$.subscribe(spy);
      de.triggerEventHandler('keyup.escape', null);

      expect(spy).toHaveBeenCalledWith({ searchBy: '' });
    });
  });
  it('Should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
