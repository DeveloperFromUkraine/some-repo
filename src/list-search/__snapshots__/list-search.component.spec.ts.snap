// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Search Component Should match snapshot 1`] = `
<ign-list-search
  ariaLabel={[Function String]}
  searchCriteria$={[Function BehaviorSubject]}
  searchInput={[Function ElementRef]}
>
  <mat-icon
    class="search-icon"
  >
    search
  </mat-icon>
  <mat-form-field
    floatplaceholder="never"
  >
    
    
    <input
      aria-label="List Search"
      matinput=""
      placeholder="Search"
    />
    
    
    
    
  
  </mat-form-field>
  
</ign-list-search>
`;
